T/209:  Table cell post-fixer will refresh a cell only when it is needed.

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Table cell post-fixer will refresh a cell only when it is needed. Closes #209.

---

### Additional information

* Needed by: https://github.com/ckeditor/ckeditor5-image/pull/316 but also can be merged anyway. as a stand-alone PR.

ps.: I need to add some tests for this.

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -39,3 +39,46 @@ function tableCellRefreshPostFixer( model ) {
 	return fixed;
 }
+
+// Check if a table cell in the view requires refreshing.
```
```suggestion
// // Check if the model table cell requires refreshing to be re-rendered to a proper state in the view.
```

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -39,3 +39,46 @@ function tableCellRefreshPostFixer( model ) {
+// @param {module:engine/model/element~Element} tableCell Table cell to check.
+// @param {String} type Type of change.
+function checkRefresh( tableCell, type ) {
+	// If children of a table cell were removed - refresh it.
```
```suggestion
	// If all children of a table cell were removed - refresh it.
```

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -39,3 +39,46 @@ function tableCellRefreshPostFixer( model ) {
+	const hasInnerText = children.some( child => child.is( 'text' ) );
+
+	// If a bare text node (not wrapped in a paragraph) was added - refresh it.
+	if ( hasInnerText ) {
```
Do we have a case when this situation occurs? We fix the table cell with paragraph ASAP, so I am not sure when this could happen 🤔

jodator replied:
There were tests for this.

scofalik replied:
We concluded that it might be for pasting so let's leave this `if`.

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -39,3 +39,46 @@ function tableCellRefreshPostFixer( model ) {
+	}
+
+	// For attribute change we only refresh single paragraphs as they might trigger <span> to <p> change in the view.
+	if ( type == 'attribute' ) {
```
This will be refreshing for a paragraph that gets second and next attributes. Maybe not a big deal but still something.

jodator replied:
huh, we might be more specific here

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -39,3 +39,46 @@ function tableCellRefreshPostFixer( model ) {
+	const hasInnerParagraph = children.some( child => child.is( 'paragraph' ) );
+
+	// If there is no paragraph in table cell then the view doesn't require refreshing.
+	if ( !hasInnerParagraph ) {
```
What if there was a clean paragraph (`<span>`) and it was replaced by an image?

jodator replied:
No need to refresh.

scofalik commented on tests/converters/table-cell-paragraph-post-fixer.js:
```diff
@@ -221,4 +229,58 @@ describe( 'Table cell paragraph post-fixer', () => {
+
+		postFixerSpy.resetHistory();
+
+		// Insert table row with one table cell
```
Copypaaaasta

scofalik commented on tests/converters/table-cell-paragraph-post-fixer.js:
```diff
@@ -221,4 +229,58 @@ describe( 'Table cell paragraph post-fixer', () => {
+			'</table>'
+		) );
+
+		sinon.assert.calledOnce( postFixerSpy );
```
Should "not be run" but `.calledOnce()`?

scofalik replied:
Ugh, okay, this is convoluted :|. Called once means that no other post-fixer was called. I wonder if we can spy this in a better way (`Differ#refreshItem` might be a little less bullet-proof though). :(

scofalik replied:
OTOH, `calledTwice` is not bullet-proof either.

jodator replied:
I'll check, I forgot about spying on differ :man_facepalming:

scofalik commented on tests/converters/table-cell-paragraph-post-fixer.js:
```diff
@@ -221,4 +229,58 @@ describe( 'Table cell paragraph post-fixer', () => {
+		sinon.assert.calledOnce( postFixerSpy );
+	} );
+
+	it( 'should be run on changing attribute of an paragraph in a table cell', () => {
```
```suggestion
	it( 'should be run on changing attribute of a paragraph in a table cell', () => {
```

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -39,3 +39,46 @@ function tableCellRefreshPostFixer( model ) {
+	// For other changes (insert/remove) the <span> to <p> change can occur when:
+	// - sibling is added to a single paragraph (childCount == 2)
+	// - sibling is removed and single paragraph is left (childCount == 1)
+	return tableCell.childCount < 3;
```
This will trigger a refresh when removing one of three elements, leaving with two.

scofalik commented:
Okay, we talked F2F about this and the solution with `WeakMap` has a small "initialization" problem. OTOH, there are fewer cases that we really have to worry about, so I trust the proposed solution more.

We concluded that we can go with the proposed solution but I'd like to see more tests to feel safer and maybe some additional comments in the code explaining why the `if`s that we have are enough and what assumptions we took so that it can work (for example, we were talking that in the end we only need to refresh if there is a paragraph because only existing paragraph might need that -- there is an `if` in the code but it is not explained clear enough).

jodator commented:
I've rewrite this one case and added more comments on why those checks.

Also I've updated the **proper** test file - I don't know why I've changed tests for other post-fixer :man_facepalming:.

jodator commented:
One more finding... this text wasn't suppose to be here.

scofalik commented on src/converters/table-cell-refresh-post-fixer.js:
```diff
@@ -25,17 +25,61 @@ export default function injectTableCellRefreshPostFixer( model ) {
+// @param {String} type Type of change.
+function checkRefresh( tableCell, type ) {
+	// If all children of a table cell were removed - refresh it.
+	if ( !tableCell.childCount ) {
```
Actually, now I wonder, do we have to do that? Didn't we have another post-fixer for that earlier? 🤔

scofalik replied:
What I mean is that we need to auto-paragraph this cell rather than refresh it, but refreshing it might trigger auto-paragraphing (in other post-fixer). And I just wonder if this shouldn't be elsewhere (if it is needed) or removed.

jodator replied:
yeah.. probably also

jodator replied:
fixing

jodator replied:
fixed

jodator replied:
fixed
